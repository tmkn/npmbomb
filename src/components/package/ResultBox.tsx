/** @jsx jsx */
import { jsx, css, keyframes } from "@emotion/core";
import React, { useContext } from "react";

import { mq, textColor } from "../../css";
import { ResultsTable, Num } from "../shared/results/Results";
import { Divider } from "../shared/divider/Divider";
import { scaleDuration } from "./CountUp";
import { Info } from "../shared/info/Info";
import { GuessContext } from "./GuessBox";
import { TextLink } from "../shared/link/TextLink";
import { TabView, ITab } from "../shared/tabview/TabView";
import { TreeTest } from "../shared/tree/Tree";

function plural(count: number): string {
    return count === 1 ? "dependency" : "dependencies";
}

export const Summary: React.FC = () => {
    const { package: pkg } = useContext(GuessContext);
    const SummaryTab =  (
        <React.Fragment>
            <Info>
                <span>
                    <TextLink href={`https://www.npmjs.com/package/${pkg.name}/v/${pkg.version}`}>
                        {pkg.name}@{pkg.version}
                    </TextLink>{" "}
                    defines{" "}
                    <b>
                        {pkg.directDependencies} direct {plural(pkg.directDependencies)}
                    </b>{" "}
                    which explode into{" "}
                    <b>
                        {pkg.dependencies} {plural(pkg.dependencies)} overall
                    </b>
                    , resulting in{" "}
                    <b>
                        {pkg.distinctDependencies} distinct {plural(pkg.distinctDependencies)}
                    </b>
                    .
                </span>
            </Info>
        </React.Fragment>
    );
    const tabs: ITab[] = [
        {header:`Summary`, content: SummaryTab},
        {header: `Tree`, content: <React.Fragment><TreeTest /></React.Fragment>}
    ]

    return <TabView tabs={tabs} />
};

interface IResultBoxProps {
    guess: number;
    actual: number;
    distinct: number;
}

export const ResultBox: React.FC<IResultBoxProps> = ({ guess, actual, distinct }) => {
    const fadeIn = keyframes`
        from {
            height: 0;
            opacity: 0;
        }

        to {
            height: auto;
            opacity: 1;
        }
    `;

    const containerStyle = css({
        [mq[0]]: {
            //display: "none",
            opacity: 0,
            height: 0,
            animation: `${fadeIn} 500ms ease forwards`,
            animationDelay: `${scaleDuration}ms`,
            h2: {
                marginTop: 0
            }
        }
    });

    const resultStyle = css({
        [mq[0]]: {
            marginBottom: "1rem",
            color: textColor
        }
    });

    const distance = Math.abs(guess - actual);

    return (
        <div css={containerStyle}>
            {actual > 0 && <Summary />}
            {guess !== actual && (
                <React.Fragment>
                    <h2>Results</h2>
                    <Divider margin={"1rem 0"} />
                    <ResultsTable columns={2}>
                        <div>Your Guess:</div>
                        <Num>{guess}</Num>
                        <div>Actual:</div>
                        <Num>{actual}</Num>
                    </ResultsTable>
                    <Divider margin={"1rem 0"} />
                    <div css={resultStyle}>
                        <React.Fragment>
                            You were off by <Num>{distance}</Num>
                        </React.Fragment>
                    </div>
                </React.Fragment>
            )}
        </div>
    );
};
