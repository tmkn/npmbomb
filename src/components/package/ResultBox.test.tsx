import React from "react";
import { BrowserRouter } from "react-router-dom";
import { render, fireEvent, waitForElement } from "@testing-library/react";

import { ResultBox } from "./ResultBox";

describe("<ResultBox />", () => {
    test("matches snapshot", () => {
        const { asFragment } = render(<ResultBox guess={23} actual={45} distinct={123} />);

        expect(asFragment()).toMatchSnapshot();
    });
});
