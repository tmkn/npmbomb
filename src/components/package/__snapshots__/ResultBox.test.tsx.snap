// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultBox /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-7emoiw"
  >
    <h2>
      Results
    </h2>
    <div
      class="css-14vnwjz"
      style="margin: 1rem 0px;"
    />
    <div
      class="css-60ri6o"
    >
      <div>
        Your Guess:
      </div>
      <span
        class="css-12eoq8x"
      >
        23
      </span>
      <div>
        Actual:
      </div>
      <span
        class="css-12eoq8x"
      >
        45
      </span>
      <div>
        Distinct:
      </div>
      <span
        class="css-12eoq8x"
      >
        123
      </span>
    </div>
    <div
      class="css-14vnwjz"
      style="margin: 1rem 0px;"
    />
    <div
      class="css-ziif3o"
    >
      You were off by 
      <span
        class="css-12eoq8x"
      >
        22
      </span>
    </div>
  </div>
</DocumentFragment>
`;
